### PR TITLE
enableWhenExpression follow up PR for maintaining CodeHealth

### LINKED ISSUE #1345: this api (internal api) change is fine. But what this means is that the `questionnaireResponseItemRetriever` is going to become a little redundant. I think in this case we should remove `questionnaireResponseItemRetriever` and instead implement the functionality of finding the questionnaire response item from `linkId` inside of the `enablement` package.
this api (internal api) change is fine. But what this means is that the `questionnaireResponseItemRetriever` is going to become a little redundant. I think in this case we should remove `questionnaireResponseItemRetriever` and instead implement the functionality of finding the questionnaire response item from `linkId` inside of the `enablement` package.

i'm happy for this to be done in a follow-up PR.

_Originally posted by @jingtang10 in https://github.com/google/android-fhir/pull/1294#discussion_r857097359_

### PR DESCRIPTION
**Pull request by RaaziaTarique**
**IMPORTANT: All PRs must be linked to an issue (except for extremely trivial and straightforward changes).**

Fixes #1345 

**Description**
Remove questionnaireResponseItemRetriever and instead implement the functionality of finding the questionnaire response item from linkId inside of the enablement package.

**Alternative(s) considered**
Have you considered any alternatives? And if so, why have you chosen the approach in this PR?

**Type**
Choose one: Code health

**Screenshots (if applicable)**

**Checklist**
- [x] I have read and acknowledged the [Code of conduct](https://github.com/google/android-fhir/blob/master/docs/CODE_OF_CONDUCT.md).
- [x] I have read the [Contributing](https://github.com/google/android-fhir/wiki/Contributing) page.
- [ ] I have signed the Google [Individual CLA](https://cla.developers.google.com/about/google-individual), or I am covered by my company's [Corporate CLA](https://cla.developers.google.com/about/google-corporate ).
- [x] I have discussed my proposed solution with code owners in the linked issue(s) and we have agreed upon the general approach.
- [x] I have run `./gradlew spotlessApply` and `./gradlew spotlessCheck` to check my code follows the style guide of this project.
- [x] I have run `./gradlew check` and `./gradlew connectedCheck` to test my changes locally.
- [x] I have built and run the demo app(s) to verify my change fixes the issue and/or does not break the demo app(s).

### CODE REVIEW

File: datacapture/src/main/java/com/google/android/fhir/datacapture/enablement/EnablementEvaluator.kt [hunk 1]
```diff
@@ -164,3 +150,25 @@ val fhirPathEngine: FHIRPathEngine =
   with(FhirContext.forCached(FhirVersionEnum.R4)) {
     FHIRPathEngine(HapiWorkerContext(this, DefaultProfileValidationSupport(this)))
   }
+
+/** Map from link IDs to questionnaire response items. */
+private fun linkIdToQuestionnaireResponseItemMap(questionnaireResponse: QuestionnaireResponse) =
```
Comment (maimoonak) [inline datacapture/src/main/java/com/google/android/fhir/datacapture/enablement/EnablementEvaluator.kt:155]: there is a similar method/map in QuestionnaireViewModel (linkIdToQuestionnaireResponseItemMap). We should utilize that. This map creation every time for each evaluation would be expensive. Or is there any specific reason to not use that map!

Comment (RaaziaTarique) [reply datacapture/src/main/java/com/google/android/fhir/datacapture/enablement/EnablementEvaluator.kt:155]: We were previously using the same mapping but it was being done in the QuestionnaireViewModel as we were not passing the QuestionnaireResponse to EnablementEvaluator. I can use the same method present QuestionnaireViewModel though.

**Review by maimoonak — APPROVED** [review]

**Comment by jingtang10** [comment]
@joiskash please take a look at this. We spoke about this before. This is a really important refactoring.

**Review by joiskash — COMMENTED** [review]
These changes look good, but I am unable to figure out one part:

In `questionnaireResponseItemChangedCallback` the `linkIdToQuestionnaireResponseItemMap` is modified. Is this map equivalent to the map that you get after `createLinkIdToQuestionnaireResponseItemMap` with the updated questionnaireResponse ?

**Comment by RaaziaTarique** [comment]
> These changes look good, but I am unable to figure out one part:
> 
> In `questionnaireResponseItemChangedCallback` the `linkIdToQuestionnaireResponseItemMap` is modified. Is this map equivalent to the map that you get after `createLinkIdToQuestionnaireResponseItemMap` with the updated questionnaireResponse ?
> 
> @RaaziaTarique @jingtang10

Yes, as I have moved createLinkIdToQuestionnaireResponseItemMap to `MoreQuestionnaireItemComponent` so, I modified it as per  `EnablementEvaluator`.
@joiskash   CC: @jingtang10

**Comment by Tarun-Bhardwaj** [comment]
@f-odhiambo are you aware if this PR is ready for review by Google?

**Comment by f-odhiambo** [comment]
> @f-odhiambo are you aware if this PR is ready for review by Google?

This PR is ready for review @Tarun-Bhardwaj

**Comment by jingtang10** [comment]
i'll raise a separate pr

**Pull request closed without merging**